lua/lspconfig/util.lua

```diff
@@ -298,6 +301,27 @@ function M.server_per_root_dir_manager(make_config)
         client_instance.workspace_folders = {}
       end
       table.insert(client_instance.workspace_folders, params.event.added[1])
+      if client_instance.config.new_folder_restart then
+        client_instance.stop()
+        local attached_buffers = client_instance.attached_buffers
+        new_config.workspace_folders = client_instance.workspace_folders
+        local id = lsp.start_client(new_config)
+        for buf in pairs(attached_buffers) do
+          lsp.buf_attach_client(buf, id)
+        end
+        --remove id from cache
+        for _, client_ids in pairs(clients) do
+          for k, v in pairs(client_ids) do
+            if v == client_instance.id then
+              table.remove(client_ids, k)
+              table.insert(client_ids, id)
```

Comment:
this is a lot of new code. should live in a separate private function, e.g. `do_restart` that is called from here.